lib/evmone/eof.cpp

```diff
@@ -129,45 +430,117 @@ std::variant<EOF1Header, EOFValidationError> validate_eof1(
         return *error;
 
     const auto& section_headers = std::get<EOFSectionHeaders>(section_headers_or_error);
-    EOF1Header header{section_headers[CODE_SECTION], section_headers[DATA_SECTION]};
+    const auto& code_sizes = section_headers[CODE_SECTION];
+    const auto data_size =
+        section_headers[DATA_SECTION].empty() ? uint16_t{0} : section_headers[DATA_SECTION][0];
 
-    const auto error_instr =
-        validate_instructions(rev, {&container[header.code_begin()], header.code_size});
-    if (error_instr != EOFValidationError::success)
-        return error_instr;
+    const auto header_size = eof_header_size(section_headers);
 
-    return header;
-}
+    const auto types_or_error =
+        validate_types(container, header_size, section_headers[TYPE_SECTION]);
+    if (const auto* error = std::get_if<EOFValidationError>(&types_or_error))
+        return *error;
+    const auto& types = std::get<std::vector<EOF1TypeHeader>>(types_or_error);
 
-}  // namespace
+    std::vector<uint16_t> code_offsets;
+    const auto type_section_size =
+        section_headers[TYPE_SECTION].empty() ? 0u : section_headers[TYPE_SECTION][0];
+    auto offset = header_size + type_section_size;
+    for (const auto code_size : code_sizes)
+    {
+        code_offsets.emplace_back(static_cast<uint16_t>(offset));
```

Comment:
For static casts like these, should we add an assertion prior that the value fits it?